### PR TITLE
Exact match multiclass

### LINKED ISSUE #1332: Add ExactMatch metric for multiclass
### Motivation

Add `MulticlassExactmatch` to classification: https://torchmetrics.readthedocs.io/en/stable/classification/exact_match.html
Based on discussion here
https://pytorch-lightning.slack.com/archives/C010R4A7189/p1668192787385079
where a user wants sequence/sentence level accuracy where all tokens (multiclass) need to match to count as an match.

I think we had the feature before the refactor (through the `subset_accuracy` argument in `accuracy` metric) but it got lost in the refactor at some point. My mistake.

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes #1332 

## Before submitting

- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/torchmetrics/functional/classification/exact_match.py [hunk 1]
```diff
@@ -18,14 +18,57 @@
...
+    preds: Tensor,
+    target: Tensor,
+    num_classes: int,
+    multidim_average: Literal["global", "samplewise"] = "global",
+    ignore_index: Optional[int] = None,
+    validate_args: bool = True,
+) -> Tensor:
+    top_k, average, = (
+        1,
+        None,
```
Comment (Borda) [inline src/torchmetrics/functional/classification/exact_match.py:61]:
```suggestion
        None
```

File: src/torchmetrics/classification/exact_match.py [hunk 2]
```diff
@@ -160,5 +258,36 @@ def update(self, preds: Tensor, target: Tensor) -> None:  # type: ignore
...
+
+
+class ExactMatch:
+    r"""Computes Exact match (also known as subset accuracy). Exact Match is a stricter version of accuracy where
+    all labels have to match exactly for the sample to be correctly classified.
+
+    This module is a simple wrapper to get the task specific versions of this metric, which is done by setting the
+    ``task`` argument to either ``'multiclass'`` or ``multilabel``. See the documentation of
+    :mod:`MulticlassExactMatch` and :mod:`MultilabelExactMatch` for the specific details of
+    each argument influence and examples.
```
Comment (Borda) [inline src/torchmetrics/classification/exact_match.py:272]: can add simple doctest just for sick the code is tested...

Comment (Borda) [reply src/torchmetrics/classification/exact_match.py:272]:
```suggestion
    each argument influence and examples.
        Legacy Example:
        >>> target = torch.tensor([[[0, 1], [2, 1], [0, 2]], [[1, 1], [2, 0], [1, 2]]])
        >>> preds = torch.tensor([[[0, 1], [2, 1], [0, 2]], [[2, 2], [2, 1], [1, 0]]])
        >>> metric = ExactMatch(task="multiclass", num_classes=3, multidim_average='global')
        >>> metric(preds, target)
        tensor(0.5000)
        
    Legacy Example:
        >>> target = torch.tensor([[[0, 1], [2, 1], [0, 2]], [[1, 1], [2, 0], [1, 2]]])
        >>> preds = torch.tensor([[[0, 1], [2, 1], [0, 2]], [[2, 2], [2, 1], [1, 0]]])
        >>> metric = ExactMatch(task="multiclass", num_classes=3, multidim_average='samplewise')
        >>> metric(preds, target)
        tensor([1., 0.])
```

File: src/torchmetrics/functional/classification/exact_match.py [hunk 3]
```diff
@@ -129,5 +203,35 @@ def multilabel_exact_match(
...
+    validate_args: bool = True,
+) -> Tensor:
+    r"""Computes Exact match (also known as subset accuracy). Exact Match is a stricter version of accuracy where
+    all classes/labels have to match exactly for the sample to be correctly classified.
+
+    This function is a simple wrapper to get the task specific versions of this metric, which is done by setting the
+    ``task`` argument to either ``'multiclass'`` or ``'multilabel'``. See the documentation of
+    :func:`multiclass_exact_match` and :func:`multilabel_exact_match` for the specific details of
+    each argument influence and examples.
+    """
```
Comment (Borda) [inline src/torchmetrics/functional/classification/exact_match.py:228]:
```suggestion
    Legacy Example:
        >>> target = torch.tensor([[[0, 1], [2, 1], [0, 2]], [[1, 1], [2, 0], [1, 2]]])
        >>> preds = torch.tensor([[[0, 1], [2, 1], [0, 2]], [[2, 2], [2, 1], [1, 0]]])
        >>> multiclass_exact_match(preds, target, task="multiclass", num_classes=3, multidim_average='global')
        tensor(0.5000)
        
        >>> target = torch.tensor([[[0, 1], [2, 1], [0, 2]], [[1, 1], [2, 0], [1, 2]]])
        >>> preds = torch.tensor([[[0, 1], [2, 1], [0, 2]], [[2, 2], [2, 1], [1, 0]]])
        >>> multiclass_exact_match(preds, target, task="multiclass", num_classes=3, multidim_average='samplewise')
        tensor([1., 0.])
    """
```

**Review by Borda — APPROVED** [review]

**Review by justusschock — APPROVED** [review]
